### ISSUE TITLE
Package-Requires not respected by emacsPackages breaks emacs packages at runtime

### ISSUE THREAD

**Issue by ParetoOptimalDev** [NixOS/nixpkgs#182645]
### Describe the bug

Doesn't this make nix managed emacs packages way less likely to work than those installed with package install or straight.el? I guess they are buildable and reproducible in that sense, but not reproducible in terms of expected behavior which is arguably what matters most.

org-ql requires `org-super-agenda 1.2` but `org-super-agenda 1.3-pre` is installed.

<details>
<summary>Package-Requires line from org-ql</summary>

```
Package-Requires: ((emacs "26.1") (dash "2.18.1") (f "0.17.2") (map "2.1") (org "9.0") (org-super-agenda "1.2") (ov "1.0.6") (peg "1.0") (s "1.12.0") (transient "0.1") (ts "0.2-pre"))
```
</summary>
</details>
So we can see the Package-Requires line states it wants `org-super-agenda 1.2`. Is that important?

For this package, yes. For instance the maintainer says this to a Guix user in alphapapa/org-ql#166:

> Upon closer inspection, I see that the two UNSAFE failures are due to the fact that you are using emacs-org-super-agenda@1.1.1-3.a87ca11. **org-ql 0.5 explicitly depends on org-super-agenda 1.2**, because that version has the necessary support to operate these features safely. That's why I released org-super-agenda 1.2 before releasing org-ql 0.5.

<details>
<summary>showing org-super-agenda-1.3-pre is used</summary>

```
nix-repl> pkgs.emacsPackages.org-ql.propagatedBuildInputs                                     [ «derivation /nix/store/s5g203km2yg1j00fhm72vrjph2725f1j-emacs-dash-20220607.1100.drv» «derivation /nix/store/6w7cf3agmljxzci7zwv1hypmqxl13vr4-emacs-f-20220608.943.drv» «derivation /nix/store/539c2swnba4l97ymf9lkj0304vy61spj-emacs-map-3.2.1.drv» «derivation /nix/store/ca57fawaprlkgzc0b0vgx1v0qdp55r6z-emacs-org-9.5.4.drv» «derivation /nix/store/sxw9wyn6kvfdbr7i7ypj7v2cw56xixvy-emacs-org-super-agenda-20210928.916.drv» «derivation /nix/store/g9xvmr6f5kk9hpahwi1pp92wvl9w8fcd-emacs-ov-20200326.1042.drv» «derivation /nix/store/kp2yw2jhnd55k7h4l2ia023hs37l93zm-emacs-peg-1.0.drv» «derivation /nix/store/k7fidbpjn747m8iyv93ywi6lgjpkvirl-emacs-s-20210616.619.drv» «derivation /nix/store/2xvdw2scn1vpan0l3727wy9pjcpcv68y-emacs-transient-20220527.2213.drv» «derivation /nix/store/19mdk6fx0k1cwyjmj47s7q94virk8j81-emacs-ts-20210813.1617.drv» ]
```

Note that we are looking for `-emacs-org-super-agenda-20210928.916` and finding it's version. Since the default emacs Package is used this tells us the version:

```
nix-repl> pkgs.emacsPackages.org-super-agenda.version     
"20210928.916"
```

But that's not the right version, so let's look up the outPath of the source and in the org-super-agenda.el file directly:

```
$ grep Version /nix/store/4h7ks5l2magvik2nypri5ks60mdypwrs-source//org-super-agenda.el 
;; Version: 1.3-pre
```
</details>

### Steps To Reproduce
1. Add the `emacsPackages.org-ql` package
2. Open emacs, find-library RET org-ql RET, see Package-Requires wants `org-super-agenda 1.2=
3. `find-library RET org-super-agenda RET` look for `Version` and see that `org-super-agenda 1.3-pre` is installed

### Expected behavior

Emacs `Package-requires` to be respected, because otherwise it's a roll of the dice determining whether emacs packages actually work or not when installed via nix.

### Notify maintainers

I don't think there's a specific single maintainer for this, so I'll attempt to notify who it seems is in charge of or understands the most about the nix emacs infrastructure.

@ttuegel - #27083 seems related to this issue, but specifying melpaPackages, melpaStable, etc doesn't change any package versions in this case and doesn't address the core issue
@rycee - Probable related work on #82604
@adisbladis - created auto-updating emacs overlay, likely has knowledge about this issue
@Apteryks - Updated the Guix infrastructure documention to emphasize that Package-requires matter
@vlaci - I think nix-doom-emacs accounts for Package-Requires so you might have knowledge here

If others know more relevant people I should tag, please also tag them in the comments.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
$ nix-shell -p nix-info --run "nix-info -m"
this path will be fetched (0.06 MiB download, 0.30 MiB unpacked):
  /nix/store/01bkry44wndk7m5d1bi3gyfmc3izs9lz-bash-interactive-5.1-p16-dev
copying path '/nix/store/01bkry44wndk7m5d1bi3gyfmc3izs9lz-bash-interactive-5.1-p16-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.53, NixOS, 22.11 (Raccoon), 22.11.20220715.65fae65`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.10.2`
 - channels(pareto-optimal-dev): `"home-manager, nixos-21.03pre261529.d9dba88d08a, nixos-20.09-20.09.2497.4a75ca4a4e7"`
 - nixpkgs: `/nix/store/l4ysisamwr0jdf6w2za4zygj9ajlf0wb-source`
```

**Comment by alphapapa** [comment]
As I wrote in the issue you mentioned this one in: In Emacs, the Package-Requires header specifies minimum versions that are required, not exact versions. It would be incorrect to forbid later versions from being used.

**Comment by vlaci** [comment]
Nix-doom-emacs doesn't take this into account, it just blindly collects the packages requested from straight.el to install and resolves them from emacs-overlay

**Comment by ParetoOptimalDev** [comment]
> As I wrote in the issue you mentioned this one in: In Emacs, the Package-Requires header specifies minimum versions that are required, not exact versions. It would be incorrect to forbid later versions from being used, and it would likely introduce unsolvable conflicts between packages.
> 
> You quoted me as saying this to a user:
> 
> > Upon closer inspection, I see that the two UNSAFE failures are due to the fact that you are using emacs-org-super-agenda@1.1.1-3.a87ca11. org-ql 0.5 explicitly depends on org-super-agenda 1.2, because that version has the necessary support to operate these features safely. That's why I released org-super-agenda 1.2 before releasing org-ql 0.5.
> 
> As I wrote, the needed version was 1.2, but that user was using 1.1. It would be fine if the user were using 1.3-pre, because it's a later version than 1.2.

@alphapapa Thanks for the clarification. I thought the package requires were like a lockfile.

Closing this since current behavior of selecting the latest emacs package is correct.
